Implement `keep_empty` for `unnest_longer()`

DavisVaughan opened the pull request:
Closes https://github.com/tidyverse/tidyr/issues/1363
Closes https://github.com/tidyverse/tidyr/issues/1339

We are going to use `keep_empty = FALSE` as the default to be consistent with `unnest()`, `unchop()`, `separate_longer_position()`, etc. This is a breaking change as previously `NULL` elements implicitly had `keep_empty = TRUE` applied to them. However, we are now consistent with how we handle `NULL` and empty vectors like `integer()`, which I think is important.

DavisVaughan commented on R/unnest-longer.R:
```diff
@@ -20,38 +22,41 @@
 #'   column names. The default, `NULL`, gives the output columns the same names
 #'   as the input columns.
 #' @param indices_to A string giving the column name (or names) to store the
-#'   the inner names or positions (if not named) of the values. If multiple
```
I just fixed a typo of duplicated `the the` split over these two lines. No other changes besides a reflow of the comment

DavisVaughan commented on R/unnest-longer.R:
```diff
@@ -194,21 +205,23 @@ col_to_long <- function(col,
 # Convert a list element to a long tibble with:
 # - 1 col (2 if `indices_include` is `TRUE`)
 # - N rows, where `N = vec_size(x)`
+#   - If `keep_empty == TRUE` and `vec_size(x) == 0`, then `N = 1` instead
 elt_to_long <- function(x,
                         index,
                         name,
                         values_to,
                         indices_to,
                         indices_include,
+                        keep_empty,
                         error_call = caller_env()) {
   if (is.null(x)) {
-    x <- unspecified(1L)
+    x <- unspecified()
 
     if (indices_include) {
       if (is.integer(index)) {
-        index <- 1L
+        index <- integer()
       } else {
-        index <- NA_character_
+        index <- character()
```
General idea is to have 1 standard path we always apply to `NULL`, and then if `keep_empty = TRUE` then the other path below handles both the updated `NULL` result we created here and the empty vector case

DavisVaughan commented on R/unnest-longer.R:
```diff
@@ -220,6 +233,21 @@ elt_to_long <- function(x,
+      if (is.integer(index)) {
+        index <- 1L
+      } else {
+        index <- NA_character_
```
I am slightly questioning our decision to use `NA` when there aren't any names over `""`, like what `names2()` would return

DavisVaughan replied:
I'm really referring to this

``` r
tibble(
  x = list(1:2, c(a = 3, b = 4))
) %>%
  unnest_longer(x)
#> # A tibble: 4 × 2
#>       x x_id 
#>   <dbl> <chr>
#> 1     1 <NA> 
#> 2     2 <NA> 
#> 3     3 a    
#> 4     4 b
```

DavisVaughan replied:
Maybe we could change ^ to use `""`, and then consistently use `NA_integer_` or `NA_character_` when we promote empty vectors to size 1? So its like:

No names anywhere:
- Positional names for existing values
- `NA_integer_` if a zero length vector was promoted to size 1

At least one name somewhere:
- Elements with names use their names
- Elements without names use `""`
- `NA_character_` if a zero length vector was promoted to size 1

DavisVaughan replied:
Using `""` for values without names is probably more consistent in this case

``` r
tibble(
  x = list(c(a = 1, 2), c(3, 4))
) %>%
  unnest_longer(x)
#> # A tibble: 4 × 2
#>       x x_id 
#>   <dbl> <chr>
#> 1     1 "a"  
#> 2     2 ""   
#> 3     3 <NA> 
#> 4     4 <NA>
```

hadley replied:
I kind of like using `""` for unnamed elements and `NA` for elements that didn't exist. But I can see how that might be slicing things too fine.

DavisVaughan replied:
Isn't that what I'm suggesting?

hadley replied:
Like this case:

``` r
library(tidyr)
tibble(x = list(c(a = 1, 2), c(3, 4), NULL)) %>%
  unnest_longer(x, keep_empty = TRUE)
#> # A tibble: 5 × 2
#>       x x_id 
#>   <dbl> <chr>
#> 1     1 "a"  
#> 2     2 ""   
#> 3     3 <NA>  # should be ""
#> 4     4 <NA>  # should be ""
#> 5    NA <NA> # should be NA
```

<sup>Created on 2023-01-05 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

DavisVaughan commented on revdep/cran.md:
```diff
@@ -1,9 +1,9 @@
 
- * We saw 6 new problems
- * We failed to check 11 packages
+ * We saw 4 new problems
```
2 were false positives from before, so that's good

hadley commented on tests/testthat/test-unnest-longer.R:
```diff
@@ -173,22 +256,39 @@ test_that("can unnest multiple columns (#740)", {
+  out <- unnest_longer(df, c(a, b), keep_empty = TRUE, indices_include = TRUE)
+
+  expect_identical(out$a, c(NA, NA, 3L, 4L))
+  expect_identical(out$a_id, c(1L, 1L, 1L, 2L))
```
These values confused me for a while because I though they must be out indexes; but they're inner indexes, some of which point to elements that don't exist. (`NULL[[1]]` is ok by coincidence, but `double()[[1]]` is not).

DavisVaughan replied:
These are now `NA` instead of `1L`

DavisVaughan commented:
This PR grew a bit, but I realized that I could vectorize `col_to_long()` a bit more by:
- Moving all names handling to `collect_indices_info()`
- Adding `list_init_null()` to handle converting `NULL`s to either size 0 or size 1 vectors
- Adding `list_init_typed()` to follow that up and convert size 0 vectors to size 1 vectors

We already have `list_init_empty()` that tries to do both of these at once, and we use it in a few other places. It isn't quite right for our needs here, so I had to break it apart into its two functions.

After this PR I'll go back and remove `list_init_empty()` in favor of using these two helpers.

I'm also confident that when I do that I will be able to fix https://github.com/tidyverse/tidyr/issues/1327 based on the way I've implemented these two helpers.

---

I've also implemented the changes to the name handling we discussed here https://github.com/tidyverse/tidyr/pull/1442#discussion_r1062645387

DavisVaughan commented on R/unnest-longer.R:
```diff
@@ -141,97 +149,89 @@ col_to_long <- function(col,
-      indices_to = indices_to,
-      indices_include = indices_include,
-      error_call = error_call
+  if (!list_all_vectors2(col)) {
+    cli::cli_abort(
+      "List-column {.var {name}} must contain only vectors or `NULL`.",
+      call = error_call
     )
```
Nicely able to pull this check out of the iteration helper

DavisVaughan commented on R/unnest-longer.R:
```diff
@@ -141,97 +149,89 @@ col_to_long <- function(col,
-        index <- NA_character_
-      }
-    }
+  info <- list_init_null(col, sizes, ptype = ptype, size = size_null)
+  col <- info$x
+  sizes <- info$sizes
+
+  if (keep_empty) {
+    info <- list_init_typed(col, sizes, ptype = ptype, size = 1L)
+    col <- info$x
+    sizes <- info$sizes
   }
```
These two vectorized helpers now handle most of what `elt_to_long()` was doing

DavisVaughan commented on R/unnest-longer.R:
```diff
@@ -141,97 +149,89 @@ col_to_long <- function(col,
+    ptype <- new_long_indexed_frame(ptype, index_ptype, 0L, names)
+    col <- pmap(
+      list(col, indices, sizes),
+      function(elt, index, size) new_long_indexed_frame(elt, index, size, names)
+    )
   } else {
-    out <- list(x)
-    names(out) <- values_to
+    name <- values_to
+    ptype <- new_long_frame(ptype, 0L, name)
+    col <- map2(
+      col,
+      sizes,
+      function(elt, size) new_long_frame(elt, size, name)
+    )
```
The `map2()` and `pmap()` paths are now the most expensive part of `unnest_longer()`. Even though the functions are fairly simple (wrap into a data frame), if `col` is long the repeated calls to `.f` just naturally take awhile.

DavisVaughan commented on R/unnest-longer.R:
```diff
@@ -261,10 +259,20 @@ collect_indices_info <- function(col, indices_include) {
     # Indices are requested, and some elements are named.
     # For any unnamed elements, generate `NA` indices.
-    indices[unnamed] <- map(sizes[unnamed], vec_rep, x = NA_character_)
+    indices[unnamed] <- map(sizes[unnamed], vec_rep, x = "")
```
Here is the change to `""` for fully unnamed vectors

DavisVaughan commented on R/unnest-longer.R:
```diff
@@ -261,10 +259,20 @@ collect_indices_info <- function(col, indices_include) {
+    empty <- sizes == 0L
+
+    if (any(empty)) {
+      # `NULL` or typed empty elements get an `NA` index of the right type
+      index_empty <- vec_init(index_ptype)
+      indices[empty] <- list(index_empty)
```
And here is the change to consistently use `NA` of the right type for `NULL` or typed empty elements that we "keep" (rather than using a mix of `1L` and `NA_character_` somewhat inconsistently)

DavisVaughan commented on R/utils.R:
```diff
@@ -171,6 +172,122 @@ list_init_empty <- function(x,
+  if (!vec_is_list(x)) {
+    cli::cli_abort("`x` must be a list.", .internal = TRUE)
+  }
+  if (is_list_of(x)) {
+    cli::cli_abort("`x` can't be a list-of. Unclass first and provide `ptype`.", .internal = TRUE)
+  }
```
The restriction of "no list-of" isn't technically necessary, in particular the previous `list_init_empty()` helper tried to allow list-ofs and use their information. But this is part of what caused the failure in https://github.com/tidyverse/tidyr/issues/1327, and it is cleaner to extract out the `ptype` of list-ofs ahead of time, strip them down to bare lists, and pass all that through instead.

hadley replied:
This makes sense to me because `list_of()` can't contain `NULL`s, right?

DavisVaughan replied:
They can actually (that's the missing value) but can't contain `unspecified()` (although that branch probably wouldn't be hit for list-ofs because you'd typically also be supplying `ptype`)

hadley commented on R/utils.R:
```diff
@@ -171,6 +172,122 @@ list_init_empty <- function(x,
+#' have the correct size where `NULL` elements were replaced.
+#'
+#' @noRd
+list_init_null <- function(x, sizes, ..., ptype = NULL, size = 1L) {
```
The name made me thing think would initialise a list of `NULL`s. Maybe `list_replace_null()`?

hadley commented on R/utils.R:
```diff
@@ -171,6 +172,122 @@ list_init_empty <- function(x,
+
+  if (vec_any_missing(x)) {
+    null <- vec_detect_missing(x)
+    null <- which(null)
```
Can't you assume that `vec_assign()` will do this?

DavisVaughan replied:
It definitely will. But since I use `null` in two separate assignments below, it seemed slightly more efficient to `which()` it once here
